core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
I'm reluctant to make `rdds.zipWithIndex` parallel by default with no way to disable it. It's a common case for RDDs in the union to wrap a Hadoop InputFormat. I don't think it's normally an expectation that getSplits could be called concurrently and custom InputFormats are fairly common. I know Parquet/Hive had a bug where Parquet's InputFormat was caching results, so I think users should be able to disable this to avoid similar issues.

How about:
1. Separate the flag from the partition eval method so it is clear that the flag indicates whether partitions are calculated in parallel (for testing)
2. Use the default thread pool instead of custom parallelism

Would that work?